src/DomainHost.sol

```diff
@@ -0,0 +1,445 @@
+    }
+
+    // --- MCD Support ---
+
+    /// @notice Set the global debt ceiling for the remote domain
+    /// @dev Please note that pre-mint DAI cannot be removed from the remote domain
+    /// until the remote domain signals that it is safe to do so
+    /// @param wad The new debt ceiling [WAD]
+    function _lift(uint256 wad) internal auth returns (uint256 _rid) {
+        require(vat.live() == 1, "DomainHost/vat-not-live");
+
+        uint256 rad = wad * RAY;
+        uint256 dlineWad;
+        int256 dline = _int256(rad) - _int256(line);
+
+        if (dline > 0) {
+            // We are issuing new pre-minted DAI
+            dlineWad = uint256(dline) / RAY;                    // No precision loss as line is always a multiple of RAY
+            vat.slip(ilk, address(this), int256(dlineWad));     // No need for conversion check as amt is under a RAY of full size
+            vat.frob(ilk, address(this), address(this), address(this), int256(dlineWad), int256(dlineWad));
+            daiJoin.exit(escrow, dlineWad);
+
+            grain += dlineWad;
+        }
+
+        line = rad;
+
+        _rid = rid++;
+
+        emit Lift(wad);
+    }
+
+    /// @notice Withdraw pre-mint DAI from the remote domain
+    /// @param _lid Local ordering id
+    /// @param wad The amount of DAI to release [WAD]
+    function _release(uint256 _lid, uint256 wad) internal ordered(_lid) {
+        require(vat.live() == 1, "DomainHost/vat-not-live");
+
+        // Fix any permissionless repays that may have occurred
+        (uint256 ink, uint256 art) = vat.urns(ilk, address(this));
+        if (art < ink) {
+            address _vow = vow;
+            uint256 diff = ink - art;
+            vat.suck(_vow, _vow, diff * RAY); // This needs to be done to make sure we can deduct sin[vow] and vice in the next call
+            vat.grab(ilk, address(this), address(this), _vow, 0, _int256(diff));
+        }
+
+        require(dai.transferFrom(escrow, address(this), wad), "DomainHost/transfer-failed");
+        daiJoin.join(address(this), wad);
+        int256 repay = -_int256(wad);
+        vat.frob(ilk, address(this), address(this), address(this), repay, repay);
+        vat.slip(ilk, address(this), repay);
+
+        grain -= wad;
+
+        emit Release(wad);
+    }
+
+    /// @notice Guest is pushing a surplus (or deficit)
+    /// @param _lid Local ordering id
+    /// @param wad The amount of DAI to push (or pull) [WAD]
+    function _push(uint256 _lid, int256 wad) internal ordered(_lid) {
+        if (wad >= 0) {
+            dai.transferFrom(address(escrow), address(this), uint256(wad));
+            daiJoin.join(address(vow), uint256(wad));
+        } else {
+            sin += uint256(-wad);
+        }
+
+        emit Push(wad);
+    }
+
+    /// @notice Move bad debt from the remote domain into the local vow
+    /// @dev This is a potentially dangerous operation as a malicious domain can drain the entire surplus buffer
```

Comment:
To be precise, can mint max DAI, so much worse than draining the surplus buffer.